### ISSUE TITLE
Dynamic Form post not working

### ISSUE THREAD

**Issue by johnwason** [OrchardCMS/Orchard#5778]
I have been trying to implement this example:

http://www.ideliverable.com/blog/customizing-user-registration-and-login-with-dynamic-forms-and-workflows

but the workflow does not receive the event and the model validation does not get called.  I do not think the post controller is being called. I have tried other dynamic forms and the workflow is not called.

**Comment by remesq** [comment]
I presume this is similar to the issue found here: https://github.com/OrchardCMS/Orchard/issues/5765.

Since this was listed as the "bug" issue, I will post my findings here and link the other issue to this.

I compared my 1.9.0 install with the 1.9.1 install I was upgrading to and discovered that the same site in 1.9.0 works when you click "Submit" (i.e., if the form is empty or has errors, the page refreshes with the client validation error. In 1.9.1 that same site does not work and the problem I mentioned in https://github.com/OrchardCMS/Orchard/issues/5765 occurs.

I did a quick file compare of my 1.9.0 and 1.9.1 folders, and while there are some changes in most of the files, the one that may be the culprit is the ~/Controllers/FormController.cs file. I could be wrong (not a programmer), but specifically the 1.9.1 files Submit action:

```
[HttpPost]
public ActionResult Submit(int contentId, string formName) {
    var layoutPart = _layoutManager.GetLayout(contentId);
    var form = _formService.FindForm(layoutPart, formName);
    var urlReferrer = Request.UrlReferrer != null && Url.IsLocalUrl(Request.UrlReferrer.ToString()) ? Request.UrlReferrer.ToString() : "~/";

    if (form == null) {
        Logger.Warning("The specified form \"{0}\" could not be found.", formName);
        _notifier.Warning(T("The specified form \"{0}\" could not be found."));
        return Redirect(urlReferrer);
    }

    var values = _formService.SubmitForm(layoutPart, form, ValueProvider, ModelState, this);
    this.TransferFormSubmission(form, values);

    if (!ModelState.IsValid)
        return Redirect(urlReferrer);

    if(Response.IsRequestBeingRedirected)
        return new EmptyResult();

    var redirectUrl = !String.IsNullOrWhiteSpace(form.RedirectUrl) ? _tokenizer.Replace(form.RedirectUrl, new { Content = layoutPart.ContentItem }) : urlReferrer;
    return Redirect(redirectUrl);
}
```

I imagine this might be where the problem lies.

**Comment by johnwason** [comment]
I set a breakpoint in FormController.Submit and it doesn't get called.  I think the problem is in one of the view models.

**Comment by remesq** [comment]
I can confirm that I changed the original code in `~/Controllers/FormController.cs` and it now works as it should. Sorry if I am mistaken and this is not exactly your problem. It sounded similar.

In the `Submit` action, I changed this (1.9.1):

`var urlReferrer = Request.UrlReferrer != null && Url.IsLocalUrl(Request.UrlReferrer.ToString()) ? Request.UrlReferrer.ToString() : "~/";`

to the following (which is what is in 1.9.0):

`var urlReferrer = Request.UrlReferrer != null ? Request.UrlReferrer.ToString() : "~/";`

Of course, this does not necessarily mean it is the solution. `!= null && Url.IsLocalUrl` is there for a reason. Not sure what. I did read up that `IsLocalUrl` will evaluate to TRUE on a relative path (i.e., `/Orchard/SomePath` but FALSE on an absolute path (i.e., `http://www.mysite.com/Orchard/SomePath`), so maybe this is the problem?

Is it the condition (`? :`) or is it something else that needs to be done?

Some references:
[StackOverflow Question 1](http://stackoverflow.com/questions/8633782/why-does-url-islocalurl-return-false-if-the-url-contains-a-fragment)
[StackOverflow Question 2] (http://stackoverflow.com/questions/25522008/returnurl-and-islocalurl?lq=1)

Sorry if I am misunderstanding the issue.

**Comment by johnwason** [comment]
That change fixed the form for my site. Did you specify a target URL in your form?  I left it blank.  I think the goal was to prevent cross-site attacks by making sure the incoming request is from the local website, however Url.IsLocalUrl is not smart enough to check the hostname.  It just checks for the local URL, and it is possible that the UrlReferrer was returning a relative URL during the original development instead of the complete URL with hostname that is more typical.  The author should have returned an error here instead of just silently failing. 

My forms have the action

<form action="/OrchardLocal/Orchard.DynamicForms/Form/Submit" method="POST" enctype="multipart/form-data">

that seems to work. What is your test form returning for the action?

**Comment by remesq** [comment]
Sorry about that. Was messing around with settings and looking at tokens on the form and accidently saved a redirect when I did not mean to. Thanks.

But, yes, that change does work. The problem is I am not sure what problems taking out `!= null && Url.IsLocalUrl` might cause. But for now it works.

**Comment by sebastienros** [comment]
It seems @sfmskywalker fixed it with this commit c8f2c2731170aef6b54f44896ac5bb277a69c05d
on 1.9.x

**Comment by remesq** [comment]
Is 1.9.x staging for 1.9.2? Just wanted clarification if this fix was going to make it for the 1.9.2 version.

Thanks, in advance.
